signer/src/wsts_state_machine.rs

```diff
@@ -30,33 +39,435 @@ use wsts::traits::Signer as _;
 /// hash bytes while for the signing rounds we identify the state machine
 /// by the sighash bytes.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct StateMachineId([u8; 32]);
+pub enum StateMachineId {
+    /// Identifier for a DKG state machines
+    Dkg(model::BitcoinBlockHash),
+    /// Identifier for a Bitcoin signing state machines
+    BitcoinSign(SigHash),
+    /// Identifier for a rotate key verification signing round
+    RotateKey(PublicKeyXOnly, model::BitcoinBlockHash),
+}
 
 impl From<&model::BitcoinBlockHash> for StateMachineId {
     fn from(value: &model::BitcoinBlockHash) -> Self {
-        StateMachineId(value.to_byte_array())
+        StateMachineId::Dkg(*value)
     }
 }
 
 impl From<SigHash> for StateMachineId {
     fn from(value: SigHash) -> Self {
-        StateMachineId(value.to_byte_array())
+        StateMachineId::BitcoinSign(value)
+    }
+}
+
+/// A trait for converting a message into another type.
+pub trait FromMessage {
+    /// Convert the given message into the implementing type.
+    fn from_message(message: &Message) -> Self
+    where
+        Self: Sized;
+}
+
+impl FromMessage for Packet {
+    fn from_message(message: &Message) -> Self {
+        Packet {
+            msg: message.clone(),
+            sig: Vec::new(),
+        }
+    }
+}
+
+/// Wrapper for a WSTS FIRE coordinator state machine.
+#[derive(Debug, Clone, PartialEq)]
+pub struct FireCoordinator(fire::Coordinator<Aggregator>);
+
+impl std::ops::Deref for FireCoordinator {
+    type Target = fire::Coordinator<Aggregator>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for FireCoordinator {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Wrapper for a WSTS FROST coordinator state machine.
+#[derive(Debug, Clone, PartialEq)]
+pub struct FrostCoordinator(frost::Coordinator<Aggregator>);
+
+impl std::ops::Deref for FrostCoordinator {
+    type Target = frost::Coordinator<Aggregator>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for FrostCoordinator {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// A trait for WSTS state machines.
+pub trait WstsCoordinator
+where
+    Self: Sized,
+{
+    /// Creates a new coordinator state machine.
+    fn new<I>(signers: I, threshold: u16, message_private_key: PrivateKey) -> Self
+    where
+        I: IntoIterator<Item = PublicKey>;
+
+    /// Gets the coordinator configuration.
+    fn get_config(&self) -> Config;
+
+    /// Save the state required to reconstruct the state machine.
+    fn save(&self) -> SavedState;
```

Comment:
Why do have/need this? It seems we call it in the coordinator, but we do nothing with the returned state